numpy/core/src/multiarray/arrayobject.c

```diff
@@ -385,26 +475,51 @@ array_dealloc(PyArrayObject *self)
         PyObject_ClearWeakRefs((PyObject *)self);
     }
     if (fa->base) {
-        /*
-         * UPDATEIFCOPY means that base points to an
-         * array that should be updated with the contents
-         * of this array upon destruction.
-         * fa->base->flags must have been WRITEABLE
-         * (checked previously) and it was locked here
-         * thus, unlock it.
-         */
-        if (fa->flags & NPY_ARRAY_UPDATEIFCOPY) {
-            PyArray_ENABLEFLAGS(((PyArrayObject *)fa->base),
-                                                    NPY_ARRAY_WRITEABLE);
-            Py_INCREF(self); /* hold on to self in next call */
-            if (PyArray_CopyAnyInto((PyArrayObject *)fa->base, self) < 0) {
+        int retval;
+        if (PyArray_FLAGS(self) & NPY_ARRAY_WRITEBACKIFCOPY)
+        {
+            char * msg = "WRITEBACKIFCOPY requires a call to "
+                "PyArray_ResolveWritebackIfCopy or "
+                "PyArray_DiscardWritebackIfCopy before array_dealloc is "
+                "called.";
+            /* 2017-Nov-10 1.14 */
+            if (DEPRECATE(msg) < 0) {
+                /* dealloc must not raise an error, best effort try to write
+                   to stderr and clear the error
+                */
+                PyObject * s;
+#if PY_MAJOR_VERSION < 3
+                s = PyString_FromString(msg);
+#else
+                s = PyUnicode_FromString(msg);
+#endif
+                if (s) {
+                    PyErr_WriteUnraisable(s);
```

Comment:
This is not correct usage - the argument should be the _context_ - the message is already printed along with the exception type. Right now, the message is printed twice.

Off the top of my head, `(PyObject *)&PyArray_Type` would be a safe context to pass. In theory you could pass `self`, but calling `array_repr` sounds a little too risky.

if nothing else, passing `NULL` would cause the message alone to be printed.

Reply:
If we used the py3 `tp_finalize` slot, then we could pass `ndarray.__del__` as the context.

Reply:
> Right now, the message is printed twice

This code should only run if ``DEPRECATE`` (which is [``PyErr_WarnEx``](https://docs.python.org/3/c-api/exceptions.html#issuing-warnings) ) raises an exeption, in which case the exception will be ignored since this is a deallocator. In order to be sure the error is emitted, I wanted to print it to ``stderr``, and the documentation seems to indicate that is what [``PyErr_WriteUnraisable``](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WriteUnraisable) does. Should I print to stderr and then emit ``PyErr_WriteUnraisable`` as well, or simply skip it altogether? Since there was no example in NumPy I looked at how ``PyErr_WriteUnraisable`` is used in Pandas as a hint.

Reply:
> the documentation seems to indicate that is what PyErr_WriteUnraisable does

Correct, but it does this by looking at `sys,exc_info`, not by looking at its argument

Reply:
I think I understand, the entire ``if()`` clause should be replaced by the one-liner 
```
PyErr_WriteUnraisable((PyObject *)&PyArray_Type);
```
How should I proceed since this pull request has been merged? Issue another pull request based off HEAD?

Reply:
hmm, it seems cpython prints both the context from ``PyErr_Fetch`` and ``repr(obj)`` for a call to ``PyErr_WriteUnrasiable`` but PyPy only prints the ``repr(obj)`` without the context. I will file an issue there

Reply:
I'd personally go for a new PR based on the last commit of this PR, but going off head would work fine too.